### PR TITLE
Deployment script

### PR DESCRIPTION
**Pull request by samwfowler**
This update should add a deployment script to create all resources required in AWS for lambda function to run every 5 minutes with read/write access to an s3 ingestion-bucket and access to cloudwatch for logging.

Currently a test script is deployed to lambda and logs to cloudwatch every 5 minutes, this is in place of the data extraction script.

### CODE REVIEW

File: deployment/deploy_ingestion.sh [hunk 1]
```diff
@@ -0,0 +1,112 @@
+echo "Loading available resource variables from resource.json..."
+SUFFIX=$(cat resources.json | jq .resource_suffix | tr -d '"')
+S3_READ_WRITE_POLICY=$(cat resources.json | jq .s3_read_write_policy | tr -d '"')
+CLOUDWATCH_POLICY=$(cat resources.json | jq .cloudwatch_log_policy | tr -d '"')
+IAM_ROLE=$(cat resources.json | jq .iam_role | tr -d '"')
+EVENTBRIDGE_RULE=$(cat resources.json | jq .eventbridge_rule | tr -d '"')
```
Comment (AnthonyWongGit) [inline deployment/deploy_ingestion.sh:6]: fghfgh

File: deployment/deploy_ingestion.sh [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+fi
+FUNCTION_NAME=data-ingestion-${SUFFIX}
+AWS_ACCOUNT_ID=$(aws sts get-caller-identity | jq .Account | tr -d '"')
+CODE_BUCKET_NAME=code-bucket-${SUFFIX}
+INGESTION_BUCKET=ingested-data-bucket-1
+wait
+
+echo "Generating AWS secret for source database credentials..."
+aws secretsmanager create-secret --name source-db-creds --secret-string file://__source_creds.json
+aws secretsmanager create-secret --name destination-db-creds --secret-string file://__destination_creds.json
```
Comment (Hana008) [inline deployment/deploy_ingestion.sh]: Add to readme instructions to create relevant files (source_creds.json and destination_creds.json) in the expected format

File: deployment/deploy_ingestion.sh [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+wait
+
+echo "Creating bucket code bucket '${CODE_BUCKET_NAME}'..."
+aws s3 mb s3://${CODE_BUCKET_NAME}
+wait
+
+echo "Creating function deployment package..."
+cd ../src/ingestion-folder/
+zip ../../deployment/__test-ingestion.zip __test-ingestion.py
```
Comment (Hana008) [inline deployment/deploy_ingestion.sh]: Received error that '__test-ingestion.py' does not match any existing files, looks to be due to mismatched file names

File: deployment/deploy_ingestion.sh [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+wait
+
+echo "Creating function deployment package..."
+cd ../src/ingestion-folder/
+zip ../../deployment/__test-ingestion.zip __test-ingestion.py 
+cd ../../deployment
+wait
+
+echo "Uploading the deployment package..."
+aws s3 cp test-ingestion.zip s3://${CODE_BUCKET_NAME}/${FUNCTION_NAME}/test-ingestion.zip
```
Comment (Hana008) [inline deployment/deploy_ingestion.sh]: Received error 'The user-provided path test-ingestion.zip does not exist.', looks to be due to mismatched zip names

File: deployment/deploy_ingestion.sh [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    RESOURCES_JSON=$(cat resources.json)
+    jq --arg policy "${EVENTBRIDGE_RULE}" '.eventbridge_rule |= $policy' <<< $RESOURCES_JSON > resources.json
+else
+    echo "Eventbridge rule already created, skipping..."
+fi
+wait
+
+echo "Giving permission to eventbridge rule to invoke lambda function..."
+aws lambda add-permission --function-name ${FUNCTION_NAME} --principal events.amazonaws.com --statement-id eventbridge-invoke-${SUFFIX} --action "lambda:InvokeFunction" --source-arn ${EVENTBRIDGE_RULE}
```
Comment (Hana008) [inline deployment/deploy_ingestion.sh]: This command gives output to the console, requires user to manually quit in order to continue. Recommend outputting output to /dev/null instead.
Screenshot 2022-12-16 at 11 53 44

File: deployment/deploy_ingestion.sh [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+fi
+wait
+
+echo "Giving permission to eventbridge rule to invoke lambda function..."
+aws lambda add-permission --function-name ${FUNCTION_NAME} --principal events.amazonaws.com --statement-id eventbridge-invoke-${SUFFIX} --action "lambda:InvokeFunction" --source-arn ${EVENTBRIDGE_RULE} 
+wait
+
+TARGET_JSON=$(jq --arg aws_id "${AWS_ACCOUNT_ID}" --arg func_name "${FUNCTION_NAME}" '.Arn |= "arn:aws:lambda:us-east-1:" + $aws_id + ":function:" + $func_name' templates/eventbridge_targets.json)
+echo "Adding lambda function as target for eventbridge rule..."
+aws events put-targets --rule ingestion-rule-${SUFFIX} --targets "[${TARGET_JSON}]"
```
Comment (Hana008) [inline deployment/deploy_ingestion.sh]: This command gives output to the console, requires user to manually quit in order to continue. Recommend outputting output to /dev/null instead.
Screenshot 2022-12-16 at 11 54 12

**Review by Hana008 — CHANGES_REQUESTED** [review]
Requested some changes as comments throughout the code. As previously discussed, separate out into individual files and add testing. Looks good overall!

File: deployment/deploy_ingestion.sh (hunk 2 above)
Comment (samwfowler) [reply deployment/deploy_ingestion.sh]: Added section into readme re. pre-requisites for running deployment script

File: deployment/deploy_ingestion.sh (hunk 3 above)
Comment (samwfowler) [reply deployment/deploy_ingestion.sh]: This was due to having the __test-ingestion.py in the .gitignore, I have renamed these and ensured the correct files are uploaded to the repo.

File: deployment/deploy_ingestion.sh (hunk 4 above)
Comment (samwfowler) [reply deployment/deploy_ingestion.sh]: This was due to the __test-ingestion.py compiling failing previously and thus not creating a test-ingestion.zip file, will not be a problem now that the file-names are updated.

Comment (vasilecondrea) [reply deployment/deploy_ingestion.sh]: we modified the __test-ingestion.py there should not be any probs now, thx!

File: deployment/deploy_ingestion.sh (hunk 5 above)
Comment (vasilecondrea) [reply deployment/deploy_ingestion.sh]: it does not block running the script to the end on our side, I guess we can look at it again if that happens again. thx

**Pull request merged**